developers/ide/karaf.md

```diff
@@ -0,0 +1,74 @@
+---
+layout: developersguide
+title: Karaf debugging
+---
+
+# Karaf debugging
+
+This procedure allows for testing and debugging while running openHAB under [Apache Karaf](https://karaf.apache.org/).
+
+### Limitations of the demo project
+
+The `org.openhab.demo.app` project is great for testing and debugging in most circumstances, but not everything behaves identically to that of a "real" openHAB installation, which uses [Apache Karaf](https://karaf.apache.org/).
+Karaf provides a runtime environment with a lot of built-in functionality that isn't available when running the demo project.
+As a consequence, some things function slightly differently or aren't available at all, such as loading of `KAR` add-ons.
+
+### Preparation
+
+If you haven't already, clone the [openhab distro repository](https://www.github.com/openhab/openhab-distro).
+
+Make sure that the bundles you want to test or debug have been installed in your local Maven repository, using `mvn clean install -pl :<bundle coordinates>` or similar. Any bundles not found in your local Maven repository will be downloaded in their latest snapshot version. If you want to test or debug an older version, you must therefore build and install everything into your local Maven repository first.  
+
+### Building
+
+From the `openhab-distro` folder, run
+```
+mvn package -Pdev
+```
+Do not use `mvn install` since it's not possible to install a folder into the local Maven repository.
+
+### Launching
+
+The result of the build can be found in `openhab-distro/distributions/openhab/target/openhab-<version>`. From this folder, you can launch openHAB under Karaf by running `start` or `start_debug`.
+The difference between the two is that `start_debug` will also allow a debugger to be attached to port `5005`.
+The console from which you did this will now become the [Karaf console](/docs/administration/console.html) and will accept any normal Karaf command.
+
+You can now open this openHAB installation in your browser via [http://localhost:8080/](http://localhost:8080/).
+
+To stop the installation, run `system:shutdown` or `logout` from the [Karaf console](/docs/administration/console.html).
+
+### Debugging
+
+Make sure that you launched openHAB using `start_debug`, and attach the debugger of your IDE to port `5005`.
+Your IDE needs to have access to the source code that you want to debug.
+
+The exact procedure for setting up remote debugging is specific to each IDE.
+
+#### Eclipse
+
+Create a new "Debug Configuration" for remote debugging:
```

Comment:
I would consider adding this section to the Ecplise guide and link to it (similarly as for IntelliJ).

Reply:
The reason I didn't do that is that I think it would be confusing to have this in the "normal" Eclipse documentation. The reason is that you don't need to do any of this to debug using Eclipse when using the "demo app". You just start it from Eclipse and start debugging. It's only when you want to connect to "a remote" debug session that you need to follow these steps - which is the situation when running Karaf independently of Eclipse.

As I understand it, IntelliJ and VScode don't have this tight integration, so that for these, this is the "normal debug procedure". Thus, as I see it, the procedure belongs in the "main guide" for those, but not for Eclipse.

Reply:
Okay, thanks for the explanation, fine then.